perfkitbenchmarker/providers/ibmcloud/ibm_api.py

```diff
@@ -0,0 +1,577 @@
+      to list available images).
+    """
+    if args and args[0] == 'login':
+      IbmAPICommand.gen = ibmcloud.IbmCloud(account=self.ibmcloud_account_id,
+                                            apikey=self.ibmcloud_apikey,
+                                            verbose=self.debug,
+                                            version='v1',
+                                            silent=True,
+                                            force=False)
+      IbmAPICommand.gen_instmgr = ibmcloud.InstanceManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_imgmgr = ibmcloud.ImageManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_fipmgr = ibmcloud.FipManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_keymgr = ibmcloud.KeyManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_subnetmgr = ibmcloud.SubnetManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_volumemgr = ibmcloud.VolumeManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_vpcmgr = ibmcloud.VPCManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_sgmgr = ibmcloud.SGManager(IbmAPICommand.gen)
+    self.args = list(args)
+    self.flags = {}
+    self.user_data = None
+    self.sgid = None
+
+  def GetToken(self):
+    """Returns user token """
+    return self.gen.GetToken()
+
+  def CreateSgRules(self):
+    """Creates some default security group rules needed for vms to communicate """
+    sgid = self.flags['sgid']
+    self.gen_sgmgr.CreateRule(sgid, 'inbound', 'ipv4', '0.0.0.0/0', 'icmp', None)
+    self.gen_sgmgr.CreateRule(sgid, 'inbound', 'ipv4', '0.0.0.0/0', 'tcp', 22)
+    self.gen_sgmgr.CreateRule(sgid, 'inbound', 'ipv4', '0.0.0.0/0', 'tcp', None,
+                              port_min=1024, port_max=50000)
+    self.gen_sgmgr.CreateRule(sgid, 'inbound', 'ipv4', '0.0.0.0/0', 'udp', None,
+                              port_min=1024, port_max=50000)
+    self.gen_sgmgr.CreateRule(sgid, 'inbound', 'ipv4', '0.0.0.0/0', 'tcp', 443)
+
+  def GetSgId(self):
+    """returns default security group id """
+    logging.info('Looking up existing default sg')
+    prefix = self.flags['prefix']
+    resp = self.gen_sgmgr.List()
+    if resp and 'security_groups' in resp:
+      for item in resp['security_groups']:
+        if item['vpc']['name'].startswith(prefix):  # find the one with matching vpc name
+          return item['id']
+    else:
+      logging.error('Failed to retrieve security group id: %s', resp)
+    return None
+
+  def CreatePrefix(self):
+    """Creates address prefix on vpc needed to attach subnets
+
+    Flags:
+      vpcid: ibm cloud vpc id.
+      zone: name of the zone within ibm network.
+      cidr: ip range that this prefix will cover.
+      name: name of this prefix.
+  
+    Returns:
+      The json representation of the created address prefix.
+    """
+    return self.gen_vpcmgr.CreatePrefix(self.flags['vpcid'], self.flags['zone'],
+                                        self.flags['cidr'], name=self.flags['name'])
+
+  def CreateSubnet(self):
```

Comment:
Feels a little odd to have this all in the IbmCloud class.  Can this be done in the caller?

Reply:
all the api calls are in there, we may restructure in future but will have to stay there for now.